students/srepking/Lesson06/mailroom4.py

```diff
@@ -0,0 +1,176 @@
+
+    if dict_donors.get(some_name) is None:
+        dict_donors[some_name] = []
+    try:
+        if int(some_amount) <= 0:
+            print('\nYou entered an invalid amount!!\n')
+            return
+    except ValueError:
+        print('\nYou entered an invalid amount!!\n')
+        return ValueError
+
+    dict_donors[some_name].append(float(some_amount))
+    print('Thank you so much for the generous gift of ${0:.2f}, {1}!'
+          .format(float(some_amount), some_name))
+    return dict_donors
+
+
+
+
+def summary_donors(some_donors) -> object:
+    """Create a new dictionary with Total, number of donations,
+    and average donation amount"""
+
+    donors_f = {some_name: [sum(donations), int(len(donations)),
+                sum(donations) / int(len(donations))]
+                for some_name, donations in some_donors.items()}
+    return donors_f
+
+
+def column_name_width(donors_f):
+    name_list = list(donors_f.keys())  # creates a list of keys
+    name_wi = 11  # Establish minimum column width
+    for i in name_list:
+        if len(i) > name_wi:
+            name_wi = (len(i))  # width of name column
+    return name_wi
+
+
+def column_total_width(donors_f):
+    tot_wi = 12
+    for name, summary in donors_f.items():
+        if len(str(summary[0])) > tot_wi:
+            # width of total column
+            tot_wi = (len(str(summary[0]))) + 3
+            # width of number of donations column
+    return tot_wi
+
+def column_average_width(donors_f):
+    ave_wi = 12
+    for name, summary in donors_f.items():
+        if len(str(summary[2])) > ave_wi:
+            # width of total column
+            ave_wi = (len(str(summary[2]))) + 3
+            # width of number of donations column
+    return ave_wi
+
+def column_number_width(donors_f):
+    num_wi = 12
+    for name, summary in donors_f.items():
+        if len(str(summary[1])) > num_wi:
+            # width of total column
+            num_wi = (len(str(summary[1]))) + 3
+            # width of number of donations column
+    return num_wi
+
+def sort_list(donors_f):
+    list_sorted = sorted(donors_f, key=donors_f.__getitem__, reverse=True)
+    return list_sorted
+
+
+def report():
+    """Return a report on all the donors"""
+    donors_f = summary_donors(donors)
+    print('\n''A summary of your donors donations:')
+    name_wi = column_name_width(donors_f)
+    tot_wi = column_total_width(donors_f)
+    num_wi = column_number_width(donors_f)
+    ave_wi = column_average_width(donors_f)
+
+    list_sorted = sort_list(donors_f)
+    # Print the Table
+    print(f"{'Donor Name':{name_wi}}| {'Total Given':^{tot_wi}}| "
+          f"{'Num Gifts':^{num_wi}}| {'Average Gift':^{ave_wi}}")
+    print(f"{'-':-^{(name_wi+tot_wi+ave_wi+num_wi+8)}}")
+    # print("-")
+    for key in list_sorted:
+        temp = donors_f[key]
+        print(f"{key:{name_wi}}${temp[0]:{tot_wi}.2f}{temp[1]:^{num_wi}}   "
+              f"${temp[2]:>{ave_wi}.2f}")
+
+    print('\n')
```

Comment:
to unit test this function, separate report generation by having a function return a list or a string, and then separate function to actually print the report.

Reply:
I am not sure I am understanding. The 'sort_list()' function is returning a list of keys to print in order, and then the 'report()' function is printing. Isn't this what you are are suggesting in this comment?

Reply:
Here is an example of what I mean:
```
def get_report_data():
    # ... some code here
    for key in list_sorted:
        rows.append(f"{key:{name_wi}}${temp[0]:{tot_wi}.2f}{temp[1]:^{num_wi}}   "
              f"${temp[2]:>{ave_wi}.2f}")
    return rows


def generate_report():
    rows = get_report_data()
    print("\n".join(rows))
```

this allows you to unit test get_report_data() function without having to deal with "how do I ensure the right data is printed?" and also separates your "UI" code from logic code. This will be an actual requirement in Lesson9 version of mailroom.. but making mailroom unit testable in this lesson is the first step in that direction.

Reply:
Sorry to keep asking question on this. Is it preferable to leave all the format in the 'generate_report' function and just pass the dictionary data from 'get_report_data'? Or is the idea to also test the formatting of the string in get_report_data()?

Reply:
Ok, I put the test with formatting included. Please let me know if you intended otherwise.
Thank you!

Reply:
From a unit test point of view - you're benefiting from just having formatting in the get_report_data() so that you're getting most code coverage.

From a real case scenario.. it's kind of a grey area.. the formatting is kind of specific to print functionality which we wanted to separate. So i'm not inclined to say towards either option, I think for simplicity here it makes sense to have formatting in `get_report_data` and `generate_report` just handles the printing.